File: backend/app/schemas/route.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import List, Optional
import uuid
from app.models.enums import HealthCondition, RecommendationType

class RouteRecommendRequest(BaseModel):
    source: str
    destination: str
    travel_datetime: datetime
    health_condition: Optional[HealthCondition] = None
    # Add other preferences if needed

class RouteWaypointSchema(BaseModel):
    latitude: float
    longitude: float
    predicted_aqi: float
    aqi_category: str
    health_risk: str
    travel_time_from_start_min: float

class RouteScoreSchema(BaseModel):
    pollution_score: float
    exposure_score: float
    travel_time_score: float
    health_score: float
    smart_route_score: float
    average_aqi: float
    maximum_aqi: float
    minimum_aqi: float = 0.0
    average_pm25: float = 0.0
    average_pm10: float = 0.0
    average_temperature: float = 0.0
    average_humidity: float = 0.0
    average_wind_speed: float = 0.0
    prediction_confidence: float = 0.0
    aqi_category_distribution: dict = {}

class RecommendedRoute(BaseModel):
    route_id: Optional[uuid.UUID] = None
    rank: int
    recommendation_reason: str
    health_recommendation_level: str
    travel_time_min: float
    distance_km: float
    scores: RouteScoreSchema
    waypoints: List[RouteWaypointSchema]
    polyline: str

class RouteRecommendResponse(BaseModel):
    best_route: RecommendedRoute
    alternative_routes: List[RecommendedRoute]

class RouteHistoryResponse(BaseModel):
    id: uuid.UUID
    start_lat: float
    start_lng: float
    end_lat: float
    end_lng: float
    total_distance_km: float
    estimated_duration_min: float
    created_at: datetime
    scores: Optional[List[RouteScoreSchema]] = None

File: frontend/src/features/routes/types/route.ts
export type HealthCondition = 'none' | 'asthma' | 'copd' | 'heart_disease' | 'elderly' | 'children' | 'pregnant';

export interface RouteRecommendRequest {
  source: string;
  destination: string;
  travel_datetime: string;
  health_condition?: HealthCondition;
}

export interface RouteWaypointSchema {
  latitude: number;
  longitude: number;
  predicted_aqi: number;
  aqi_category: string;
  health_risk: string;
  travel_time_from_start_min: number;
  city_name?: string | null;
  temperature?: number | null;
  pm25?: number | null;
}

export interface RouteScoreSchema {
  pollution_score: number;
  exposure_score: number;
  travel_time_score: number;
  health_score: number;
  smart_route_score: number;
  average_aqi: number;
  maximum_aqi: number;
  minimum_aqi: number;
  average_pm25: number;
  average_pm10: number;
  average_temperature: number;
  average_humidity: number;
  average_wind_speed: number;
  is_rainy: boolean;
  prediction_confidence: number;
  aqi_category_distribution: Record<string, number>;
}

export interface RecommendedRoute {
  route_id?: string;
  start_address?: string;
  end_address?: string;
  rank: number;
  recommendation_reason: string;
  health_recommendation_level: string;
  travel_time_min: number;
  distance_km: number;
  scores: RouteScoreSchema;
  waypoints: RouteWaypointSchema[];
  polyline: string;
}

export interface RouteRecommendResponse {
  best_route: RecommendedRoute;
  alternative_routes: RecommendedRoute[];
}

export interface RouteHistoryResponse {
  id: string;
  start_lat: number;
  start_lng: number;
  start_address?: string;
  end_lat: number;
  end_lng: number;
  end_address?: string;
  total_distance_km: number;
  estimated_duration_min: number;
  created_at: string;
  scores?: RouteScoreSchema[] | null;
  waypoints?: RouteWaypointSchema[] | null;
}

export interface DistanceMatrixRequest {
  origins: string;
  destinations: string;
}

export interface DistanceMatrixResponse {
  distance_text: string;
  distance_value: number; // meters
  duration_text: string;
  duration_value: number; // seconds
}

export interface DirectionsRequest {
  source: string;
  destination: string;
}

export interface DirectionsRoute {
  distance_text: string;
  distance_value: number;
  duration_text: string;
  duration_value: number;
  polyline: string;
  waypoints: { lat: number; lng: number }[];
}

export interface DirectionsResponse {
  routes: DirectionsRoute[];
}
